### DISCUSSION TITLE
Unlocking the Potential: Exploring ZeppOS Development and Exciting Previews

### DISCUSSION THREAD

**Discussion by XiaomaiTX** [zepp-health/zeppos-forum#191] [Blog]
## Notice

> Chinese version is available : [释放潜力： 浅谈ZeppOS的开发经验及期望](https://blog.uuu4.cn/archives/39/)

## ⏩ Foreword

Hello everyone👋, I'm **XiaomaiTX**, a developer who has been in contact with ZeppOS for nearly a year. I'm glad to have the opportunity to share my development experience in ZeppOS over the past year and my expectations for the new version of ZeppOS😃.

This article is **long** and will share the content of ZeppOS and related ecology from the perspective of the **developer**. It is recommended to read it patiently. Readers can also quickly find the part they want to read through the table of contents. 

**I hope everyone can vote for this article❤️. Your support is my greatest motivation**🎉🎉🎉!

ZeppOS is an excellent watch operating system based on `FreeRTOS`. In a word, it has excellent system capabilities, perfect document support, complete tool chain support, active community support, and dedicated help from officials. Next, I will share my experience from the above points and bring you some amazing project previews and feature previews that ZeppOS3.0 will be coming🎉🎉🎉. Please read patiently!

### 📄 Perfect document support

The first is the **perfect document support**. From the first time developers came into contact with ZeppOS, ZeppOS documents have very **friendly entry support**, so that novice developers can quickly join the development of ZeppOS. The document introduces in detail how to get started, introduces ZeppOS devices and related ecology, and the detailed interface document also provides great convenience for developers. At the same time, officials also actively maintain and improve the document reading experience, which is extremely important for the ZeppOS ecology.

![Perfect document support](https://github.com/zepp-health/zeppos-forum/assets/59403296/2c5f9f7f-0f2b-4ddd-b377-70c1fee268d9)

### 🔧 Complete tool chain support

The second is **complete tool chain support**. In the process of ZeppOS development, developers will enjoy the excellent experience brought by ZeppOS toolchains such as `Zeus CLI`, `complete ts development tips`, `convenient nodejs environment installation`, `simulator`, etc. These tools greatly improve development efficiency, and developers can form a good workflow from the deployment environment to the development project. From the development and debugging of an application to the official launch of an application, **Zepp Team** will also assist developers in reviewing and uploading.

![Complete tool chain support](https://github.com/zepp-health/zeppos-forum/assets/59403296/6d7e1e5c-6201-485e-b15a-60ac4a3a45d7)

### 🤵 Strong help from the community and ZeppOS Team

In the process of developing software for ZeopOS, there is not only support from the **community**, but also **engineers and officials from ZeppOS** will do their best to help. They are active in the ZeppOS community, answering questions or helping developers. For some complex and difficult problems, ZeppOS Team will also create a private channel to be followed up and solved by **professional** engineers. I can often get help from the ZeppOS Team during the development process, which has greatly improved my development efficiency and improved my application.

![Strong help from the community and ZeppOS Team](https://github.com/zepp-health/zeppos-forum/assets/59403296/1166abbd-0a8c-45e1-a3c0-a304eb751740)

### Excellent system capabilities and official support are very important

ZeppOS based on `FreeRTOS` brings many possibilities to Huami's watches, including very **long-lasting battery life**, **rich functions**, and an **open and safe application ecosystem**.

I think excellent human-computer interaction experience is the core point of smart wearable devices. Huami's ZeppOS achieves an **excellent**, **smooth** and **user-friendly** user experience. For developers, rich system capabilities and friendly development workflows will bring rich application ecology and more interesting applications to ZeppOS. ZeppOS has prepared developers with including but not limited to very rich system capabilities such as **sensors**, **UI drawing**, **timers**, **background services**, etc., as well as official software for `breathing`, `heart rate`, `exercise`, `health`, etc.

![Amazfit GTR 4 in Zepp](https://github.com/zepp-health/zeppos-forum/assets/59403296/7a041f3b-a8b7-47c9-8476-a365899817e8)

![WatchFace](https://github.com/zepp-health/zeppos-forum/assets/59403296/d8f0de4e-010d-4a00-a82e-0446f27bb144)

In the old ZeppOS 1.0, I saw the potential for the development of ZeppOS. Even if it was not so excellent at that time, it told developers that open and perfect support is the advantage in the smart wearable operating system market. I have developed some interesting applications and third-party libraries in this version of ZeppOS, such as `Breathful`, an application that helps users with respiratory training, and `Fx libraries` that can provide smooth and beautiful non-linear animation support for application development.

After a period of time, ZeppOS Team released ZeppOS2.0 after hard development. This version of the system fixes the vulnerabilities in v1.0 and adds rich new functions. ZeppOS Team has also further improved the documentation and tool chain support. Developers With the help of richer system capabilities, we can develop more applications that meet users' expectations on ZeppOS.

### 🤨 Disadvantages of ZeppOS

Even if ZeppOS has many advantages, ZeppOS is still a project in its infancy. I still find some problems during the development process, such as:

 - ZeppOS can still provide more system capabilities for developers to use, such as Text To Speech, Speech To Text, independent Wi-Fi interface, real-time microphone interface, NFC abilities, etc., which will give developers more choices to implement their own ideas.

- Even though ZeppOS already has some communities, it still lacks an effective application feedback channel, making it difficult for users to provide effective feedback with application developers. It is recommended to open the comment system and scoring system of the application mall, or establish other feedback channels.

- For developers, there is a lack of many key statistics, such as the user's feedback experience, the types of popular applications and the user's expectations. These functions should be provided to developers by ZeppOS Team, which is very helpful for developers.

I hope to continue to improve the experience of using the development tool chain. At present, the simulator still has many shortcomings, and Zeus Cli also has some minor problems.

The purpose of raising these questions here is to help ZeppOS Team better improve and improve ZeppOS and bring better experience to users. I hope that ZeppOS Team can pay attention to it and sincerely hope that the ZeppOS ecology can better develop.

## 😎 Upcoming wonderful projects

The upcoming ZeppOS 3.0 can stimulate the potential of developers. This version of ZeppOS brings `background services`, `positioning`, `Canvas` and more sensor interfaces, so that many interesting ideas can be realized.

I have developed some interesting projects for the new ZeppOS, which will be **open source** in the form of `MIT` on `GitHub`. Developers are welcome to use or **contribute**! You can visit my GitHub homepage for more information. Let's take a look at some of the content that will come in ZeppOS 3.0😄🎉!

### 🎉🎉🎉 ZeppOS Render Engine 🎉🎉🎉

For example, the `Canvas` interface can achieve **amazing 3D rendering effects**. Can you imagine experiencing 3D-supported games or software on a **small** watch?

With just some simple code, you can render a `camera`, `perspective`, `rich models`, and `efficiently optimized 3D scenes` on the watch!

![Render Example](https://github.com/zepp-health/zeppos-forum/assets/59403296/198a7cd3-8e0e-4f27-8b72-823ba8e4e1c0)

![Code Shot](https://github.com/zepp-health/zeppos-forum/assets/59403296/33eb5c3e-66fb-4da8-8f73-8ab9b36b65c5)

Although this 3D rendering engine is very **basic**, it may bring users very interesting 3D mini-games or practical tools, allowing developers to create **wonderful** things that belong to ZeppOS!

### 🎉🎉🎉 ZeppOS Advanced Gestures  🎉🎉🎉

This is an advanced gesture recognition library based on **gesture chain recognition ideas**.

You can listen to **specific**, **single** or **multiple** controls at the same time, or set up **multiple** listeners.

You can call yourself in a gesture **callback event** to realize the recognition of complex gestures (for example, sliding left before right can trigger a specific event)

You can customize the threshold of the response condition (for example, the `long press` condition can be set to 5s), and different events can set specific thresholds.

![Example GIF](https://github.com/zepp-health/zeppos-forum/assets/59403296/a12ddada-d8b1-42ba-bf51-792e7148af1e)

![Code Shot](https://github.com/zepp-health/zeppos-forum/assets/59403296/ee920e60-af90-45fa-8bf0-19dc630ee786)

I hope that this gesture recognition library can enable developers to achieve **richer gesture recognition events**, and this library will also be open source on GitHub with MIT's open source protocol!

## 💬 Some more in-depth discussions

I learned that the underlying ZeppOS based on `FreeRTOS` is implemented by the `C/C++` language. Through the encapsulation of hardware drivers, ZeppOS Team introduced the `QuickJS` engine to interpret and run JS code. In the description of app.json in the official document, the document prompts that developers can not only use the `js interpreter` to run the applet, but also use the `C language interpreter` and `binary interpreter` to run the application. Applications written in C language or compiled into binary will be **faster** in execution efficiency than applications using JS interpreters. Is it **possible** to provide **better performance** under the same conditions?

![Runtime Options](https://github.com/zepp-health/zeppos-forum/assets/59403296/cc1da505-96d6-4ace-99ae-a1ae1167a323)

As far as I know, ZeppOS Team does not intend to expose the system capabilities of ZeppOS at the C language level, so my idea is to write some **high-performance** code in C language and load it into js code through a `third-party library` to run. I am also testing the feasibility of this method.

## 😉 Postscript

The above are some of my experiences and thoughts in the development process of ZeppOS. I hope it will be helpful to readers, and **I also hope that ZeppOS can do better and better**! 

Finally, if you think this article is useful, **you can vote for my article**❤️! **This is a great affirmation for me**🎉🎉🎉!

**Comment by viteo** [comment]
Have you heard anything new about C interpreter?
